src/linux/socket/list_device_names_utils.rs

```diff
@@ -1,66 +1,63 @@
+    },
+    nl::{NlPayload, Nlmsghdr},
+    rtnl::Ifinfomsg,
+    types::RtBuffer,
+};
+use std::convert::TryFrom;
 
-pub fn get_list_device_names_msg() -> Nlmsghdr<Rtm, Ifinfomsg<Ifla>> {
+pub fn get_list_device_names_msg() -> Nlmsghdr<Rtm, Ifinfomsg> {
     let infomsg = {
         let ifi_family =
             neli::consts::rtnl::RtAddrFamily::UnrecognizedVariant(libc::AF_UNSPEC as u8);
         // Arphrd::Netrom corresponds to 0. Not sure why 0 is necessary here but this is what the
         // embedded C library does.
         let ifi_type = Arphrd::Netrom;
         let ifi_index = 0;
-        let ifi_flags = vec![];
-        let rtattrs: Vec<Rtattr<Ifla, Vec<u8>>> = vec![];
-        Ifinfomsg::new(ifi_family, ifi_type, ifi_index, ifi_flags, rtattrs)
+        let ifi_flags = IffFlags::empty();
+        let rtattrs = RtBuffer::new();
+        let ifi_change = IffFlags::new(&[Iff::Up]);
+
+        Ifinfomsg::new(
+            ifi_family, ifi_type, ifi_index, ifi_flags, ifi_change, rtattrs,
+        )
     };
 
     let len = None;
     let nl_type = Rtm::Getlink;
-    let flags = vec![NlmF::Request, NlmF::Ack, NlmF::Dump];
+    let flags = NlmFFlags::new(&[NlmF::Request, NlmF::Ack, NlmF::Dump]);
     let seq = None;
     let pid = None;
     let payload = infomsg;
-    Nlmsghdr::new(len, nl_type, flags, seq, pid, payload)
+    Nlmsghdr::new(len, nl_type, flags, seq, pid, NlPayload::Payload(payload))
 }
 
 pub struct PotentialWireGuardDeviceName {
     pub ifname: Option<String>,
     pub is_wireguard: bool,
 }
 
-pub fn parse_ifinfomsg(
-    response: Nlmsghdr<Nlmsg, Ifinfomsg<Ifla>>,
-) -> Result<PotentialWireGuardDeviceName, ListDevicesError> {
-    let mut is_wireguard = false;
-    let mut ifname: Option<String> = None;
+impl TryFrom<Nlmsghdr<NlTypeWrapper, Ifinfomsg>> for PotentialWireGuardDeviceName {
+    type Error = ListDevicesError;
 
-    for attr in response.nl_payload.rtattrs {
-        match attr.rta_type {
-            Ifla::UnrecognizedVariant(libc::IFLA_LINKINFO) => {
-                let mut buf = StreamReadBuffer::new(&attr.rta_payload);
-                let linkinfo = Rtattr::<u16, Vec<u8>>::deserialize(&mut buf)?;
+    fn try_from(response: Nlmsghdr<NlTypeWrapper, Ifinfomsg>) -> Result<Self, Self::Error> {
+        let mut handle = response.get_payload()?.rtattrs.get_attr_handle();
 
-                if linkinfo.rta_type == libc::IFLA_INFO_KIND {
-                    let info_kind = parse_nla_nul_string(&linkinfo.rta_payload)?;
-                    if info_kind == crate::linux::consts::WG_GENL_NAME {
-                        is_wireguard = true;
-                    }
-                }
-            }
-            Ifla::Ifname => {
-                ifname = Some(parse_nla_nul_string(&attr.rta_payload)?);
-            }
-            _ => {}
-        };
+        Ok(PotentialWireGuardDeviceName {
+            ifname: handle.get_attr_payload_as::<String>(Ifla::Ifname).ok(),
```

Comment:
I think the idea was that the name can simply be missing (the code previously accounted for the possibility as well).

Reply:
Gotcha. I think there's still a subtle difference:

- The previous code accounted for a missing `ifname` since those are optional, but it still propagated parsing errors in the case that `ifname` is present but malformed.
- The change in this PR swallows parsing errors and treats `ifname` fields that are present as if they were missing entirely.

I can add a test for this in a followup PR.